backends/platform/3ds/osystem-events.cpp

```diff
@@ -0,0 +1,231 @@
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#define FORBIDDEN_SYMBOL_ALLOW_ALL
+#include "backends/timer/default/default-timer.h"
+#include "backends/platform/3ds/gui.h"
+#include "osystem.h"
+
+static Common::Mutex *eventMutex;
+static InputMode inputMode = MODE_DRAG;
+static aptHookCookie cookie;
+
+static void pushEventQueue(Common::Queue<Common::Event>* queue, Common::Event& event) {
+	Common::StackLock lock(*eventMutex);
+	queue->push(event);
+}
+
+static void eventThreadFunc(void* arg) {
+	OSystem_3DS* osys = (OSystem_3DS*) g_system;
+	auto eventQueue = (Common::Queue<Common::Event>*) arg;
+	
+	uint32 touchStartTime = osys->getMillis();
+	touchPosition lastTouch = {0,0};
+	bool isRightClick = false;
+	Common::Event event;
+	
+	while(!osys->exiting) {
+		do {
+			osys->delayMillis(10);
+		} while (osys->sleeping && !osys->exiting);
+		
+		hidScanInput();
+		touchPosition touch;
+		u32 held = hidKeysHeld();
+		u32 keysPressed = hidKeysDown();
+		u32 keysReleased = hidKeysUp();
+		
+		if (held & KEY_TOUCH) {
+			hidTouchRead(&touch);
+			osys->transformPoint(touch);
+
+			osys->warpMouse(touch.px, touch.py);
+			event.mouse.x = touch.px;
+			event.mouse.y = touch.py;
+			
+			if (keysPressed & KEY_TOUCH) {
+				touchStartTime = osys->getMillis();
+				isRightClick = (held & KEY_X || held & KEY_DUP);
+				if (inputMode == MODE_DRAG) {
+					event.type = isRightClick ? Common::EVENT_RBUTTONDOWN : Common::EVENT_LBUTTONDOWN;
+					pushEventQueue(eventQueue, event);
+				}
+			}
+			else if (touch.px != lastTouch.px || touch.py != lastTouch.py) {
+				event.type = Common::EVENT_MOUSEMOVE;
+				pushEventQueue(eventQueue, event);
+			}
+			
+			lastTouch = touch;
+		}
+		else if (keysReleased & KEY_TOUCH) {
+			event.mouse.x = lastTouch.px;
+			event.mouse.y = lastTouch.py;
+			printf("clicked %u, %u\n", lastTouch.px, lastTouch.py);
+			if (inputMode == MODE_DRAG) {
+				event.type = isRightClick ? Common::EVENT_RBUTTONUP : Common::EVENT_LBUTTONUP;
+				pushEventQueue(eventQueue, event);
+			}
+			else if (osys->getMillis() - touchStartTime < 200) {
+				// Process click in MODE_HOVER
+				event.type = Common::EVENT_MOUSEMOVE;
+				pushEventQueue(eventQueue, event);
+				event.type = isRightClick ? Common::EVENT_RBUTTONDOWN : Common::EVENT_LBUTTONDOWN;
+				pushEventQueue(eventQueue, event);
+				event.type = isRightClick ? Common::EVENT_RBUTTONUP : Common::EVENT_LBUTTONUP;
+				pushEventQueue(eventQueue, event);
+			}
+		}
+		if (keysPressed & KEY_R) {
```

Comment:
It might make sense to support GUI_KEYS_DIALOG, like Symbian and WinCE do, or alternatively, employ KeyMapper, instead of hardcoding the key meanings.